### DISCUSSION TITLE
[V5] How to download an image/document using flydrive.

### DISCUSSION THREAD

**Discussion by Pratikkshirsagar** [adonisjs/core#1763] [Help] [answered]
Hi,
I am trying to download the document from s3 using flydrive in adonis 5.
As adonis [4] previous versions has ```await Drive.disk('s3').getObject('example.jpg)``` method on drive api but I'm unable to find same method for flydrive, please help me out with any other way to download any image/document using flydrive.

**Comment by thetutlage** [comment, marked as answer]
Please share some code that you have tried to use in order to help you further

**Reply by Pratikkshirsagar** [reply]
Hi now I'am able to download jpg and png images using following code
```ts 
 public async _saveStreamToFile(file: Readable, pathname: string): Promise<unknown> {
    return new Promise((resolve, reject) => {
      const writer = fs.createWriteStream(pathname);

      file.pipe(writer);

      writer.on('finish', resolve);
      writer.on('error', reject);
    });
  }
```

```ts 
public async download(response: ResponseContract, fileName: string): Promise<void> {
    const path = join(Application.appRoot, '..', 'tmp');

    const { content } = await this.storage.disk('s3').getBuffer(fileName);

    const pathname = `${path}/${fileName}`;

    const readableInstanceStream = new Readable({
      read() {
        this.push(content);
        this.push(null);
      },
    });

    await this._saveStreamToFile(readableInstanceStream, pathname);

    return response.download(pathname);
  }
```

But still unable to download pdf documents can you please help me with that..?

**Reply by thetutlage** [reply]
There are a handful of issues here. 

- If you have the file as a buffer, then there is no need to first write it on the disk and then stream
- 2nd, there is no need to get the `buffer` and instead you should stream the file directly.

The following code (untested) should work

```ts
public async download(response: ResponseContract, fileName: string): Promise<void> {
    const path = join(Application.appRoot, '..', 'tmp');
    response.stream(this.storage.disk('s3').getStream(fileName))
}
```

**Reply by Pratikkshirsagar** [reply]
This works fine Thank You.
